lib/sage.ex

```diff
@@ -437,6 +447,57 @@ defmodule Sage do
   def run_async(sage, name, transaction, compensation, opts \\ []),
     do: add_stage(sage, name, build_operation!(:run_async, transaction, compensation, opts))
 
+  @doc """
+  For a given Sage S with transactions :t1 -> :t2 -> :t3, a call to `interleave(S, :name, f)`
+  will yield a saga with transactions :t1 -> :name_1 -> :t2 -> :name_2 -> :t3 -> :name_3.
+
+  This can be useful if you are trying to do a long computation and want to do something with
+  the intermediate results, such as logging or persistence.
+  
+  Note: 
+  - This isn't strict interleaving because a transaction is still appended at the end.
+  - Calling this function twice with the same name will give a `Sage.DuplicateStageError`, but
+  calling the function twice with a different name will compound its effects.
+  - Calling this function before the end of your saga definition will mean any stages you add after the
+  call to `interleave` will not have the intermediate stages added after them
+  """
+  @spec interleave(
+          sage :: t(),
+          name :: stage_name(),
+          intermediate_transaction :: intermediate_transaction(),
+          compensation :: compensation()
```

Comment:
Do we also want compensations to be reported in your use case? If so we'll need to add a type for it too or reuse `intermediate_transaction` here.

Reply:
Good question. My use case actually doesn't require compensations at all, I just added them for completeness/generality.

Reply:
I think we should add it for completeness. If there is some reporter on the saga progress, it should also be able to report compensation progress.

Reply:
So are you saying that an interleaved compensation should be passed the previously run compensation, to mirror the `intermediate_transaction`?